geode-core/src/main/java/org/apache/geode/internal/cache/TXReservationMgr.java

```diff
@@ -96,15 +84,19 @@ private void checkForConflict(TXRegionLockRequestImpl rr, IdentityArrayList loca
     }
   }
 
-  private void checkSetForConflict(TXRegionLockRequestImpl rr, Set s, Object[] keys,
-      IdentityArrayList localLocks) throws CommitConflictException {
-    for (int i = 0; i < keys.length; i++) {
-      if (s.contains(keys[i])) {
-        release(localLocks, true);
-        throw new CommitConflictException(
-            String.format(
-                "The key %s in region %s was being modified by another transaction locally.",
-                new Object[] {keys[i], rr.getRegionFullPath()}));
+  private void checkSetForConflict(TXRegionLockRequestImpl rr, Map<Object, Boolean> oldValue,
+      Map<Object, Boolean> keys, IdentityArrayList localLocks) throws CommitConflictException {
+    for (Map.Entry<Object, Boolean> e : keys.entrySet()) {
+      if (oldValue.containsKey(e.getKey())) {
+        if (oldValue.get(e.getKey())) {
```

Comment:
I think this fix only handles multiple gets/reads holding "read" locks case. It does not handle when some transactions hold the "read" locks using gets, but a put/delete operation occurs trying to hold "write" locks. In this case, the new write transaction should fail with CommitConflictException.
I think you can add some unit tests to verify this, possibly using setAfterConflictCheck() method to verify locks and conflicts.

Reply:
Do we want this behavior. Accordin to documentation, (detectReadConflicts):

This property causes read operations to succeed only when they read a consistent pre- or post-transactional state. If not consistent, Geode throws a CommitConflictException.

There is no mentioning, that due to ongoing read operation, write or update operations will be rejected.

Reply:
Let's see it from another angle:
If a transaction holds write lock (update), a new transaction trying to hold read lock (detectReadConflicts) would fail with commit conflict exception. 
Then why a transaction holds the read lock (detectReadConflicts) does not block a transaction try to get the write lock? How does the transaction ensure read a consistent pre- or post-transactional state if another transaction would change the state of an entry it holds the read lock?

Reply:
OK, as I see, it is also and current behavior. I agree with your comments. Just I think that documentation is misleading, because it is only talking about read operations success/failure.

Reply:
Yes, we should update the docs after your fix.